File: legacy/src/main/java/preponderous/viron/services/EnvironmentService.java
// Copyright (c) 2024 Preponderous Software
// MIT License

package preponderous.viron.services;

import java.util.Arrays;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import preponderous.viron.config.ServiceConfig;
import preponderous.viron.exceptions.ServiceException;
import preponderous.viron.models.Environment;

@Service
public class EnvironmentService {
    private final RestTemplateBuilder restTemplateBuilder;
    private final ServiceConfig serviceConfig;
    private final String baseUrl;

    @Autowired
    public EnvironmentService(RestTemplateBuilder restTemplateBuilder, ServiceConfig serviceConfig) {
        this.restTemplateBuilder = restTemplateBuilder;
        this.serviceConfig = serviceConfig;
        this.baseUrl = this.serviceConfig.getVironHost() + ":" + serviceConfig.getVironPort() + "/api/v1/environments";
    }

    public List<Environment> getAllEnvironments() {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment[]> response = restTemplate.exchange(baseUrl, HttpMethod.GET, null, Environment[].class);
        if (response.getStatusCode().isError()) {
            throw new ServiceException("Error getting environments");
        }
        return Arrays.asList(response.getBody());
    }

    public Environment getEnvironmentById(int id) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/{id}", HttpMethod.GET, null, Environment.class, id);
        if (response.getStatusCode().isError()) {
            throw new ServiceException("Error getting environment by id");
        }
        return response.getBody();
    }

    public Environment getEnvironmentByName(String name) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/name/{name}", HttpMethod.GET, null, Environment.class, name);
        if (response.getStatusCode().isError()) {
            throw new ServiceException("Error getting environment by name");
        }
        return response.getBody();
    }

    public Environment getEnvironmentOfEntity(int entityId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment> response = restTemplate.exchange(baseUrl + "/entity/{entityId}", HttpMethod.GET, null, Environment.class, entityId);
        if (response.getStatusCode().isError()) {
            throw new ServiceException("Error getting environment of entity");
        }
        return response.getBody();
    }

    public Environment createEnvironment(String name, int numGrids, int gridSize) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Environment> response = restTemplate.exchange(
                baseUrl + "/{name}/{numGrids}/{gridSize}",
                HttpMethod.POST,
                null,
                Environment.class,
                name,
                numGrids,
                gridSize
        );
        if (response.getStatusCode().isError()) {
            throw new ServiceException("Error creating environment");
        }
        return response.getBody();
    }

    public boolean deleteEnvironment(int id) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Void> response = restTemplate.exchange(baseUrl + "/{id}", HttpMethod.DELETE, null, Void.class, id);
        if (response.getStatusCode().isError()) {
            throw new ServiceException("Error deleting environment");
        }
        return response.getStatusCode().is2xxSuccessful();
    }

    public boolean updateEnvironmentName(int id, String name) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Void> response = restTemplate.exchange(
                baseUrl + "/{id}/name/{name}",
                HttpMethod.PATCH,
                null,
                Void.class,
                id,
                name
        );
        if (response.getStatusCode().isError()) {
            throw new ServiceException("Error updating environment name");
        }
        return response.getStatusCode().is2xxSuccessful();
    }
}

File: legacy/src/main/python/preponderous/viron/services/environmentService.py

# Copyright (c) 2024 Preponderous Software
# MIT License

import requests
from src.main.python.preponderous.viron.models.environment import Environment

class EnvironmentService:
    def __init__(self, viron_host: str, viron_port: int):
        self.viron_host = viron_host
        self.viron_port = viron_port

    def get_base_url(self) -> str:
        return f"{self.viron_host}:{self.viron_port}/api/v1/environments"

    def get_all_environments(self) -> list[Environment]:
        response = requests.get(f"{self.get_base_url()}")
        response.raise_for_status()
        return [Environment(**env) for env in response.json()]

    def get_environment_by_id(self, environment_id: int) -> Environment:
        response = requests.get(f"{self.get_base_url()}/{environment_id}")
        response.raise_for_status()
        return Environment(**response.json())

    def get_environment_by_name(self, name: str) -> Environment:
        response = requests.get(f"{self.get_base_url()}/name/{name}")
        response.raise_for_status()
        return Environment(**response.json())

    def get_environment_of_entity(self, entity_id: int) -> Environment:
        response = requests.get(f"{self.get_base_url()}/entity/{entity_id}")
        response.raise_for_status()
        return Environment(**response.json())

    def create_environment(self, name: str, num_grids: int, grid_size: int) -> Environment:
        response = requests.post(f"{self.get_base_url()}/{name}/{num_grids}/{grid_size}")
        response.raise_for_status()
        return Environment(**response.json())

    def delete_environment(self, environment_id: int) -> bool:
        response = requests.delete(f"{self.get_base_url()}/{environment_id}")
        response.raise_for_status()
        return response.status_code == 200

    def update_environment_name(self, environment_id: int, name: str) -> bool:
        response = requests.patch(f"{self.get_base_url()}/{environment_id}/name/{name}")
        response.raise_for_status()
        return response.status_code == 200